### PR TITLE
Project Beats: Sequencer refactor (part 2)

### PR DESCRIPTION
**Pull request by sanchitmalhotra126**
Switches over our system to using Sequencer classes for sequencing playback events (rather than having logic directly in interpreted code). For more details on the background/motivation behind this change, see this [PR description](https://github.com/code-dot-org/code-dot-org/pull/51531).

Apologies for the length of this PR 😅 since this switches the sequencing logic under the hood, I had to make sure that all cases were covered so that no functionality was changed (other than disabling old unused block modes). I'll walk through the major changes here:

1. In generated blockly code, replace calls to MusicView with calls to Sequencer. This is really the crux of the change; this allows us to remove any state logic we were tracking within generated code and simply call the Sequencer APIs directly. This mostly affects simple2; in other block modes, we'll be calling the stubbed out version of the Sequencer, so there won't be any functionality.
2. Move playbackEvents into Redux: now that the MusicPlayer no longer needs to track the state of playback events, we can move these directly into redux and have components read them from the store directly. For this to work, any time code is executed in MusicView (on play, on code change, on trigger), we read the events from the sequencer and dispatch an action to update the playback events in the redux store.
3. Relatedly, we can also move the highlighted blocks computation out of MusicPlayer and into Redux as a selector, because all the information we need is now available in the redux state
4. This also allows us to remove PlayerUtilsContext completely, now that playbackEvents can be read from the redux store.
5. We can also the value of lastMeasure into Redux since this is something the sequencer will now have to compute, as it has access to the most recent set of sequenced events.
6. A couple of small fixes, such as making blockId required, cleaning/fixing up some logic in Simple2Sequencer, copying the array of notes in a chord event to a separate array to prevent readonly violations, etc.
7. Finally, we can delete all the logic in MusicPlayer that's been moved to other components. MusicPlayer now doesn't need to keep track of playback events directly; it simply plays events given to it, tracks player state like current playback time, and serves as an intermediary between the view and the sample playback system (converting seconds to playhead time, etc). I was considering pulling the deletion into its own PR, but decided that it would be better to see the deletions alongside the places the code had moved to.

## Links

https://codedotorg.atlassian.net/browse/SL-676

## Testing story

Tested extensively with simple2 (all available features), and confirmed that other block modes don't crash. However; please feel free to pull down the branch and play around yourself!

### CODE REVIEW

File: apps/src/music/redux/musicRedux.ts [hunk 1]
```diff
@@ -28,6 +29,8 @@ interface MusicState {
   showInstructions: boolean;
   instructionsPosition: InstructionsPosition;
   isBeatPadShowing: boolean;
+  playbackEvents: PlaybackEvent[];
+  lastMeasure: number;
```
Comment (breville) [inline apps/src/music/redux/musicRedux.ts]: Should we start adding comments to all of these members?

File: apps/src/music/views/MusicView.jsx [hunk 2]
```diff
@@ -137,13 +145,20 @@ class UnconnectedMusicView extends React.Component {
       const libraryJson = values[0];
       this.library = new MusicLibrary(libraryJson);
 
+      if (getBlockMode() === BlockMode.SIMPLE2) {
+        this.sequencer = new Simple2Sequencer(this.library);
+      } else {
+        this.sequencer = new MusicPlayerStubSequencer();
+      }
+
```
Comment (breville) [inline apps/src/music/views/MusicView.jsx:153]: Just an idea, but we could use a factory pattern here later on?

File: apps/src/music/redux/musicRedux.ts (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/music/redux/musicRedux.ts]: Yeah good call, will add!

File: apps/src/music/views/MusicView.jsx (hunk 2 above)
Comment (sanchitmalhotra126) [reply apps/src/music/views/MusicView.jsx:153]: Yeah, that would make sense - this is something we can add when we handle other block modes?

File: apps/src/music/views/MusicView.jsx [hunk 3]
```diff
@@ -195,6 +211,15 @@ class UnconnectedMusicView extends React.Component {
     ) {
       this.musicBlocklyWorkspace.selectBlock(this.props.selectedBlockId);
     }
+
+    if (
+      !isEqual(
```
Comment (breville) [inline apps/src/music/views/MusicView.jsx]: I think we've used JSON serialization to do deep compares elsewhere in this lab.  Is this a similar situation?  Should we settle on one comparison technique?

Comment (sanchitmalhotra126) [reply apps/src/music/views/MusicView.jsx]: Fair point, I'm good to stick with JSON serialization for consistency! I know it's used in the (currently unused) tracks timeline view

Comment (sanchitmalhotra126) [reply apps/src/music/views/MusicView.jsx]: Updated to use JSON serialization

**Review by breville — APPROVED** [review]
This is fantastic!  Unifying this business logic in native code, rather than in interpreted code, sets this model up very nicely for the long term.  Great refactor based on solid guiding principles.

Slightly apprehensive that the unit tests aren't yet ported, because they were an integral part of coding the model initially, but we've both tested this a bunch, and porting those tests is next, so I think this is good to go.

**Comment by sanchitmalhotra126** [comment]
> Slightly apprehensive that the unit tests aren't yet ported, because they were an integral part of coding the model initially, but we've both tested this a bunch, and porting those tests is next, so I think this is good to go.

Agreed - I'll plan to fast follow with porting tests so we can maintain coverage

**Pull request merged**